include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,470 @@
+
+/**
+ * Check for physical presence of battery.
+ *
+ * @return Whether there is a battery physically present, but possibly
+ * in a disconnected or cut off state, or if we can't tell;
+ */
+typedef enum battery_present (*battery_hw_present)(const struct device *dev);
+
+/**
+ * Check for battery initialization status.
+ *
+ * @return zero if not initialized.
+ */
+typedef int (*board_battery_initialized)(const struct device *dev);
+
+/**
+ * Read nominal voltage battery is designed to supply.
+ *
+ * @param voltage	Destination for voltage in mW
+ * @return non-zero if error.
+ */
+typedef int (*battery_design_voltage)(const struct device *dev, int *voltage);
+
+/**
+ * Read absolute state of charge.
+ *
+ * @param percent	Destination for charge in percent
+ * @return non-zero if error.
+ */
+typedef int (*battery_state_of_charge_abs)(const struct device *dev, int *percent);
+
+/**
+ * Read battery remaining capacity.
+ *
+ * @param capacity	Destination for capacity in mAh
+ * @return non-zero if error.
+ */
+typedef int (*battery_remaining_capacity)(const struct device *dev, int *capacity);
+
+/**
+ * Read battery full charge capacity.
+ *
+ * @param capacity	Destination for capacity in mAh
+ * @return non-zero if error.
+ */
+typedef int (*battery_full_charge_capacity)(const struct device *dev, int *capacity);
+
+/**
+ * Read the nominal capacity the battery is designed to supply when new.
+ *
+ * @param capacity	Destination for capacity in mAh
+ * @return non-zero if error.
+ */
+typedef int (*battery_design_capacity)(const struct device *dev, int *capacity);
+
+/**
+ * Read time in minutes left when discharging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_empty)(const struct device *dev, int *minutes);
+
+/**
+ * Read run time in minutes left when discharging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_run_time_to_empty)(const struct device *dev, int *minutes);
+
+/**
+ * Read time in minutes left to full capacity when charging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_full)(const struct device *dev, int *minutes);
+
+/**
+ * Calculate battery time in minutes, under an assumed current.
+ *
+ * @param rate		Current to use for calculation, in mA.
+ *			If > 0, calculates charging time; if < 0, calculates
+ *			discharging time; 0 is invalid and sets minutes=0.
+ * @param minutes	Destination for calculated time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_at_rate)(const struct device *dev, int rate, int *minutes);
+
+/**
+ * Read battery charge cycle count.
+ *
+ * @param count		Destination for count.
+ * @return non-zero if error.
+ */
+typedef int (*battery_cycle_count)(const struct device *dev, int *count);
```

Comment:
Given this is an api that some code has to implement, some additional info would be welcome as to what counts as a cycle. E.g. Does it require discharge to min volts plus charge to max, or something else... e.g. if I discharge 66% to 33%, then 66% to 33% again, is that one cycle, two, or not even one?